datastore/tiered_datastore.nim

```diff
@@ -26,57 +28,94 @@ proc stores*(self: TieredDatastore): seq[Datastore] =
 
 method contains*(
   self: TieredDatastore,
-  key: Key): ?!bool {.locks: "unknown".} =
+  key: Key): Future[?!bool] {.async, locks: "unknown".} =
 
   var
     exists = false
 
   for store in self.stores:
-    exists = ? store.contains(key)
+    let
+      containsRes = await store.contains(key)
+
+    if containsRes.isErr: return failure containsRes.error.msg
+
+    exists = containsRes.get
+
     if exists: break
 
-  success exists
+  return success exists
 
 method delete*(
   self: TieredDatastore,
-  key: Key): ?!void {.locks: "unknown".} =
+  key: Key): Future[?!void] {.async, locks: "unknown".} =
+
+  var
+    pending: seq[Future[?!void]]
```

Comment:
What about shorter version:
```
let pending = await allFinished(
  self.stores.mapIt( it.delete(key) ))
```
It applies to get, put, delete.

Reply:
I thought about that, but  I was thinking we want to return `failure` if any of the put/delete ops had an error. It's not clear to me how to do that via `mapIt`.

Reply:
the full code:
```
  let pending = await allFinished(
    self.stores.mapIt( it.delete(key) ))

  for fut in pending:
    if fut.read().isErr: return fut.read()

  return success()
```

You are right - it applies only to delete and put

Reply:
Ah, I see now, I was reading `allFinished` as `allFutures`, the latter returns `Future[void]` but `allFinished` gives us the seq we want, will make the change.

Reply:
Done.